rust/kernel/chrdev.rs

```diff
@@ -20,10 +19,69 @@ use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
+/// Character device.
+///
+/// # Invariants
+///
+/// - [`self.0`] is valid and non-null.
+/// - [`(*self.0).ops`] is valid, non-null and has static lifetime.
+/// - [`(*self.0).owner`] is valid and, if non-null, has module lifetime.
+struct Cdev(*mut bindings::cdev);
+
+impl Cdev {
+    fn alloc(
+        fops: &'static bindings::file_operations,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult<Self> {
+        // SAFETY: FFI call.
+        let cdev = unsafe { bindings::cdev_alloc() };
+        if cdev.is_null() {
+            return Err(Error::ENOMEM);
+        }
+        // SAFETY: `cdev` is valid and non-null since `cdev_alloc()`
+        // returned a valid pointer which was null-checked.
+        unsafe {
+            (*cdev).ops = fops;
+            (*cdev).owner = module.0;
+        }
+        // INVARIANTS:
+        // - [`self.0`] is valid and non-null.
+        // - [`(*self.0).ops`] is valid, non-null and has static lifetime,
+        //   because it was coerced from a reference with static lifetime.
+        // - [`(*self.0).owner`] is valid and, if non-null, has module lifetime,
+        //   guaranteed by the [`ThisModule`] invariant.
+        Ok(Self(cdev))
+    }
+
+    fn add(&mut self, dev: bindings::dev_t, count: c_types::c_uint) -> KernelResult<()> {
+        // SAFETY: by the type invariants:
+        // - [`self.0`] is valid and non-null.
+        // - [`(*self.0).ops`] is valid, non-null, and has a static lifetime.
+        // - [`(*self.0).owner`] is valid and, if non-null, has module lifetime.
+        // This guarantees that [`(*self.0).ops`] and [`(*self.0).owner`] (if non-null)
+        // will live at least as long as [`self.0`].
```

Comment:
This is wrong! The whole reason we're doing this, is because our `struct cdev` instance might outlive the module, because it's a `kobject`. So the `cdev` ends up still alive, but with a `module` pointer to freed memory.

But the kernel C code probably has the same issue - I'm guessing and the `cdev` won't touch `owner` when all file handles to it are closed. In that case, we are good.

What's your opinion?

Reply:
In general, in the kernel, module removal is best effort. I don't know what is nowadays supposed to be guaranteed to work or not, we can ask.

`cdev_del` states that the ops can still be called even after it returns (that one should be fine, since they are built into the `kernel` crate rather than the modules), but does not mention the other fields. We would need to check if they dereference the `module` anywhere.

In any case, for the comment, I would mention the detail about `ops`, since it is the important one to guarantee to be available forever; and create an issue for `owner`s in general (of all kinds of devices etc.).

Reply:
By the way, I would avoid repeating all the invariants. It is more important to explain why those invariants are good enough for the `unsafe` block (i.e. the `ops` bit in particular).